From 6fd2b4291bb87a5dfd89dc2da226d27c2b514d95 Mon Sep 17 00:00:00 2001
From: Rob Clark <robclark@freedesktop.org>
Date: Fri, 29 Aug 2014 17:09:29 -0400
Subject: [PATCH] 
 10.3-state_tracker-gallium-fix-crash-with-st_renderbuffer-freedreno.patch

---
 src/gallium/drivers/freedreno/a2xx/fd2_texture.c  | 1 +
 src/gallium/drivers/freedreno/a3xx/fd3_texture.c  | 1 +
 src/gallium/drivers/freedreno/a4xx/fd4_texture.c  | 1 +
 src/gallium/drivers/freedreno/freedreno_surface.c | 1 +
 src/gallium/drivers/freedreno/freedreno_texture.c | 2 +-
 src/gallium/drivers/freedreno/freedreno_texture.h | 3 +++
 6 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/src/gallium/drivers/freedreno/a2xx/fd2_texture.c b/src/gallium/drivers/freedreno/a2xx/fd2_texture.c
index b70eaa6..0a2e3c3 100644
--- a/src/gallium/drivers/freedreno/a2xx/fd2_texture.c
+++ b/src/gallium/drivers/freedreno/a2xx/fd2_texture.c
@@ -135,6 +135,7 @@ fd2_sampler_view_create(struct pipe_context *pctx, struct pipe_resource *prsc,
 	so->base.texture = prsc;
 	so->base.reference.count = 1;
 	so->base.context = pctx;
+	so->base.sampler_view_destroy = fd_sampler_view_destroy;
 
 	so->tex_resource =  rsc;
 	so->fmt = fd2_pipe2surface(cso->format);
diff --git a/src/gallium/drivers/freedreno/a3xx/fd3_texture.c b/src/gallium/drivers/freedreno/a3xx/fd3_texture.c
index 567f6c7..4428333 100644
--- a/src/gallium/drivers/freedreno/a3xx/fd3_texture.c
+++ b/src/gallium/drivers/freedreno/a3xx/fd3_texture.c
@@ -221,6 +221,7 @@ fd3_sampler_view_create(struct pipe_context *pctx, struct pipe_resource *prsc,
 	so->base.texture = prsc;
 	so->base.reference.count = 1;
 	so->base.context = pctx;
+	so->base.sampler_view_destroy = fd_sampler_view_destroy;
 
 	so->tex_resource =  rsc;
 
diff --git a/src/gallium/drivers/freedreno/a4xx/fd4_texture.c b/src/gallium/drivers/freedreno/a4xx/fd4_texture.c
index f9b6dcc..f05c6f1 100644
--- a/src/gallium/drivers/freedreno/a4xx/fd4_texture.c
+++ b/src/gallium/drivers/freedreno/a4xx/fd4_texture.c
@@ -159,6 +159,7 @@ fd4_sampler_view_create(struct pipe_context *pctx, struct pipe_resource *prsc,
 	so->base.texture = prsc;
 	so->base.reference.count = 1;
 	so->base.context = pctx;
+	so->base.sampler_view_destroy = fd_sampler_view_destroy;
 
 	so->tex_resource =  rsc;
 
diff --git a/src/gallium/drivers/freedreno/freedreno_surface.c b/src/gallium/drivers/freedreno/freedreno_surface.c
index 250fe4b..a16f13c 100644
--- a/src/gallium/drivers/freedreno/freedreno_surface.c
+++ b/src/gallium/drivers/freedreno/freedreno_surface.c
@@ -51,6 +51,7 @@ fd_create_surface(struct pipe_context *pctx,
 		pipe_resource_reference(&psurf->texture, ptex);
 
 		psurf->context = pctx;
+		psurf->surface_destroy = fd_surface_destroy;
 		psurf->format = surf_tmpl->format;
 		psurf->width = u_minify(ptex->width0, level);
 		psurf->height = u_minify(ptex->height0, level);
diff --git a/src/gallium/drivers/freedreno/freedreno_texture.c b/src/gallium/drivers/freedreno/freedreno_texture.c
index eaa6629..40095b4 100644
--- a/src/gallium/drivers/freedreno/freedreno_texture.c
+++ b/src/gallium/drivers/freedreno/freedreno_texture.c
@@ -41,7 +41,7 @@ fd_sampler_state_delete(struct pipe_context *pctx, void *hwcso)
 	FREE(hwcso);
 }
 
-static void
+void
 fd_sampler_view_destroy(struct pipe_context *pctx,
 		struct pipe_sampler_view *view)
 {
diff --git a/src/gallium/drivers/freedreno/freedreno_texture.h b/src/gallium/drivers/freedreno/freedreno_texture.h
index 43571a9..a1007d1 100644
--- a/src/gallium/drivers/freedreno/freedreno_texture.h
+++ b/src/gallium/drivers/freedreno/freedreno_texture.h
@@ -39,6 +39,9 @@ void fd_set_sampler_views(struct pipe_context *pctx, unsigned shader,
 						  unsigned start, unsigned nr,
 						  struct pipe_sampler_view **views);
 
+void fd_sampler_view_destroy(struct pipe_context *pctx,
+		struct pipe_sampler_view *view);
+
 void fd_texture_init(struct pipe_context *pctx);
 
 #endif /* FREEDRENO_TEXTURE_H_ */
-- 
2.1.2

